enhancements/machine-health-checking.md

```diff
@@ -0,0 +1,204 @@
+
+
+### Non-Goals
+
+- Coordinating individual remediation systems. This must be driven by each system however they see fit, e.g. by annotating affected nodes with state and transition time.
+
+- Recording long-term stable history of all health-check failures or remediations.
+
+## Proposal
+
+- As a customer I only care about my apps availability so I want my cluster infrastructure to be self healing and the nodes to be remediated automatically.
+
+- As a cluster admin I want to define my unhealthy criteria as a list of node conditions https://github.com/kubernetes/node-problem-detector.
+
+- As a cluster admin I want to define different unhealthy criteria for pools of machines targeted for different workloads.
+
+- As a cluster admin I want to define different remediation strategies for different cloud environments. E.g deletion or reboot.
+
+- As a cluster admin I want to skip remediation to optionally prevent automated remediation if some threshold of nodes are unhealthy, because this situation may require more careful intervention. E.g a large number of nodes in a single zone are down because it's most likely a networking issue.
+
+Machine health checking is an integration point between node problem detection tooling and remediation strategies.
+
+We will provide a default remediation strategy - delete. It will delete a machine when the backed node meets the unhealthy criteria defined by the user. The upper level machine controller, e.g machineSet will then reconcile towards satisfying the expected number of replicas and will create a new one.
+
+### Implementation Details
+
+#### MachineHealthCheck CRD:
+- Enable watching a pool of machines (based on a label selector).
+- Enable defining an unhealthy criteria (based on a list of node conditions).
+- Enable choosing a remediation strategy.
+- Enable setting a threshold for the number of unhealthy nodes. If the current number is at or above this threshold no further remediation 
+
+
+E.g I want my worker machines in us-east-1a to be remediated when the backed node has `ready=false` condition for more than 10m.
+
+```yaml
+apiVersion: "healthchecking.openshift.io/v1beta1"
+kind: "MachineHealthCheck"
+metadata:
+  name: workers-us-east-1a
+  namespace: openshift-machine-api
+spec:
+  selector:
+    role: worker
+    zone: us-east-1a
+  remediationStrategy: delete
+  unhealthyConditions:
+    - name: Ready
+      status: false
+      timeout: 10m
+  maxUnavailable: 5
+status:
+  ExpectedHealthy: 10 // Total number of machines targeted by the label selector
+  CurrentHealthy: 10
+```
+
+TODO: how to define OR relation between unhealthyConditions, e.g `ready=false || ready=unknown`
+
+#### MachineHealthCheck controller:
+- Watch MachineHealthCheck resources and nodes.
+- Find machines backing nodes that meet the unhealthy criteria and signal remediaton systems with the chosen remediation strategy, e.g `healthchecking.openshift.io/remediation: reboot`.
```

Comment:
Who removes this label/annotation?

Reply:
Currently this is a part of the cluster-wide machine healthcheck configuration, are we planning to move this to be a per-Machine annotation?

Reply:
this is set on the API https://github.com/openshift/enhancements/pull/9/files#diff-53036ce0d8589cdbed15ed2e95076b54R100 as it's today https://github.com/openshift/machine-api-operator/blob/master/pkg/apis/healthchecking/v1alpha1/machinehealthcheck_types.go#L45 and signaled on resources meeting "problem" criteria via annotation for remediation systems to watch and react same as it's today as well.